docs/usor/intro-dp.md

```diff
@@ -229,3 +370,212 @@ abordare, scriere a sursei și modalități de tranziție.
         return 0;
     }
     ```
+
+<!-- Fourth Problem -->
+### Problema [Vacation](https://atcoder.jp/contests/dp/tasks/dp_c)
+
+Pentru această problemă, trebuie să aflăm satisfacția maximă pe care o poate obține Taro în $n$ zile, dacă nu are voie să repete o activitate de două ori la rând. 
+
+Din nou, se poate demonstra cu ușurință că o soluție de tip Greedy nu va merge, deoarece ne putem bloca alegeri mai bune ulterior cu alegerile curente. 
+
+Totodată, mai apare un element nou în rezolvarea acestor probleme, și anume faptul că vom avea nevoie de o nouă dimensiune pentru a păstra informații cu privire la ultima activitate efectuată de el, pentru a evita o situație în care alegem de două ori aceeași activitate. 
+
+Astfel, vom defini $dp[i][j]$ ca fiind suma maximă a satisfacției dacă am parcurs primele $i$ zile, iar ultima activitate a fost de tipul $j$, $j$ fiind $0$, $1$ sau $2$, în funcție de activitatea aleasă. 
+
+Pentru a calcula $dp[i][j]$, va trebui să ne raportăm la sumele din ziua precedentă, corespunzătoare celorlalte două activități deoarece nu avem voie să alegem aceeași activitate iar.
+
+$$
+dp[i][j] = max(dp[i-1][x]) + v[j], \forall x \neq j
+$$
+
+Din nou ca la celelalte probleme, puteți găsi mai jos abordarea recursivă și cea iterativă a problemei.
+
+!!! note "Observație"
+    Deoarece avem nevoie doar de valorile din ziua precedentă, nu este necesar să păstrăm în memorie toată matricea, ci doar ultimele două linii, optimizare pe care o vom prezenta în detaliu mai jos.
+    
+=== "Recursiv"
+
+    ```cpp
+    #include <iostream>
+    #include <vector>
+
+    using namespace std;
+
+    int solve(const int poz, const vector<vector<int>> &v, vector<vector<int>> &dp, const int activ) {
+        if (poz == 1) {
+            return v[poz][activ];
+        }
+        if (dp[poz][activ] != -1) {
+            return dp[poz][activ];
+        }
+        dp[poz][activ] = 0;
+        for (int i = 0; i < 3; i++) {
+            if (i != activ) {
+                dp[poz][activ] = max(dp[poz][activ], solve(poz-1, v, dp, i));
+            }
+        }
+        dp[poz][activ] += v[poz][activ];
+        return dp[poz][activ];
+    }
+
+    int main() {
+        int n;
+        cin >> n;
+        
+        vector<vector<int>> v(n+1, vector<int>(3));
+        vector<vector<int>> dp(n+1, vector<int>(3));
+        
+        for (int i = 1; i <= n; i++) {
+            cin >> v[i][0] >> v[i][1] >> v[i][2];
+            dp[i][0] = dp[i][1] = dp[i][2] = -1;
+        }
+        
+        cout << max(solve(n, v, dp, 0), max(solve(n, v, dp, 1), solve(n, v, dp, 2))) << '\n';
+        return 0;
+    }
```

Comment:
Ai putea face main-ul așa în schimb:
```cpp
int main() {
  int n;
  cin >> n;
        
  vector<vector<int>> v(n+1, vector<int>(3));
  vector<vector<int>> dp(n+1, vector<int>(3, -1));
        
  for (int i = 1; i <= n; i++) {
    cin >> v[i][0] >> v[i][1] >> v[i][2];
  }

  int solution = 0;
  for (int i = 0; i < 3; i++) {
    solution = max(solution, solve(n, v, dp, i));
  }
  
  cout << solution << '\n';
  return 0;
}
```

Reply:
Am schimbat și printarea de sol cu
```cpp
        cout << max({solve(n, v, dp, 0), solve(n, v, dp, 1), solve(n, v, dp, 2)}) << '\n';
```
mi s-a părut mai ok să o fac așa.

Și inițializarea vectorului, deși în practică n-aș fi de acord cu ea decât dacă am scrie clean code, deoarece cred că adăugăm constantei de la complexitatea timpului utilizând constructorul cu fill.

Reply:
> Am schimbat și printarea de sol cu
> 
> ```c++
>         cout << max({solve(n, v, dp, 0), solve(n, v, dp, 1), solve(n, v, dp, 2)}) << '\n';
> ```
> 
> mi s-a părut mai ok să o fac așa.

Mare schimbare faci de la un maxim cu 3 elemente la un for de 3 elemente care face maximul. Tot sugerez să folosești acel for, e mai extensibil. Sau cel puțin să indentezi acele acolade ca să fie clar 

> Și inițializarea vectorului, deși în practică n-aș fi de acord cu ea decât dacă am scrie clean code, deoarece cred că adăugăm constantei de la complexitatea timpului utilizând constructorul cu fill.

Și dacă faci inițializarea în 2 nivele de for-uri, ieși fix la fel, deci scoate partea aia și bag-o în constructor. Nu adaugi nicio constantă, totul e cu `memset` care e cam cea mai eficientă metodă (în funcție de optimizare, folosește și SIMD și deci se ocupă de 8/16 elemente dintr-o dată, nu o să poți întrece vreodată asta). Tu scrie cod curat și lasă compilatorul să se ocupe de STL, cei care implementează nu-s proști.